File: sdks/typescript/src/resources/collections.ts
import type { RequestClient } from "../core.js";
import { BigRAGRealtimeConnection } from "../realtime.js";
import type {
  AnalyticsResponse,
  Collection,
  CollectionListOptions,
  CollectionListResponse,
  CollectionRealtimeTokenResponse,
  CollectionStatsResponse,
  CreateCollectionBody,
  ProgressEvent,
  StatusResponse,
  UpdateCollectionBody,
} from "../types/index.js";

export class CollectionsResource {
  constructor(private readonly _client: RequestClient) {}

  list(options?: CollectionListOptions): Promise<CollectionListResponse> {
    const params: Record<string, string> = {};
    if (options?.name) params.name = options.name;
    if (options?.limit !== undefined) params.limit = String(options.limit);
    if (options?.offset !== undefined) params.offset = String(options.offset);
    if (options?.include_total !== undefined) {
      params.include_total = options.include_total ? "true" : "false";
    }
    return this._client._request("GET", "/v1/collections", { params });
  }

  async *listAll(options?: Omit<CollectionListOptions, "offset">): AsyncGenerator<Collection> {
    const pageSize = options?.limit ?? 100;
    let offset = 0;
    while (true) {
      const page = await this.list({ ...options, limit: pageSize, offset });
      for (const c of page.collections) yield c;
      if (page.collections.length < pageSize) return;
      offset += page.collections.length;
      if (page.total !== null && offset >= page.total) return;
    }
  }

  get(name: string): Promise<Collection> {
    return this._client._request("GET", `/v1/collections/${encodeURIComponent(name)}`);
  }

  create(body: CreateCollectionBody): Promise<Collection> {
    return this._client._request("POST", "/v1/collections", { json: body });
  }

  update(name: string, body: UpdateCollectionBody): Promise<Collection> {
    return this._client._request("PUT", `/v1/collections/${encodeURIComponent(name)}`, {
      json: body,
    });
  }

  delete(name: string): Promise<StatusResponse> {
    return this._client._request("DELETE", `/v1/collections/${encodeURIComponent(name)}`);
  }

  stats(name: string): Promise<CollectionStatsResponse> {
    return this._client._request("GET", `/v1/collections/${encodeURIComponent(name)}/stats`);
  }

  truncate(name: string): Promise<StatusResponse> {
    return this._client._request("POST", `/v1/collections/${encodeURIComponent(name)}/truncate`);
  }

  analytics(name: string): Promise<AnalyticsResponse> {
    return this._client._request("GET", `/v1/collections/${encodeURIComponent(name)}/analytics`);
  }

  createRealtimeToken(name: string): Promise<CollectionRealtimeTokenResponse> {
    return this._client._request(
      "POST",
      `/v1/collections/${encodeURIComponent(name)}/realtime-token`,
    );
  }

  async *streamEvents(
    name: string,
    options: { token?: string } = {},
  ): AsyncGenerator<ProgressEvent> {
    const connection = new BigRAGRealtimeConnection(this._client);
    try {
      for await (const message of connection.subscribe<ProgressEvent>("collection.events", {
        collection: name,
        token: options.token,
      })) {
        if (message.type === "event") yield message.payload;
        if (message.type === "error") throw new Error(message.message);
        if (message.type === "complete") return;
      }
    } finally {
      await connection.close();
    }
  }
}

File: sdks/rust/src/resources/collections.rs
use crate::client::RagComputer;
use crate::core::urlencode;
use crate::error::RagComputerError;
use crate::sse::SseStream;
use crate::types::collections::{
    Collection, CollectionListOptions, CollectionListResponse, CollectionStatsResponse,
    CreateCollectionBody, UpdateCollectionBody,
};
use crate::types::common::StatusResponse;

/// Collections resource — manage rag.computer collections.
pub struct Collections<'a> {
    pub(crate) client: &'a RagComputer,
}

impl Collections<'_> {
    /// List collections with optional filters.
    pub async fn list(
        &self,
        options: Option<CollectionListOptions>,
    ) -> Result<CollectionListResponse, RagComputerError> {
        let mut query = Vec::new();
        if let Some(opts) = options {
            if let Some(name) = opts.name {
                query.push(("name".into(), name));
            }
            if let Some(limit) = opts.limit {
                query.push(("limit".into(), limit.to_string()));
            }
            if let Some(offset) = opts.offset {
                query.push(("offset".into(), offset.to_string()));
            }
        }
        self.client.transport.get("/v1/collections", query).await
    }

    /// Get a collection by name.
    pub async fn get(&self, name: &str) -> Result<Collection, RagComputerError> {
        let path = format!("/v1/collections/{}", urlencode(name));
        self.client.transport.get(&path, vec![]).await
    }

    /// Create a new collection.
    pub async fn create(&self, body: CreateCollectionBody) -> Result<Collection, RagComputerError> {
        self.client.transport.post("/v1/collections", &body).await
    }

    /// Update a collection.
    pub async fn update(
        &self,
        name: &str,
        body: UpdateCollectionBody,
    ) -> Result<Collection, RagComputerError> {
        let path = format!("/v1/collections/{}", urlencode(name));
        self.client.transport.put(&path, &body).await
    }

    /// Delete a collection and all its data.
    pub async fn delete(&self, name: &str) -> Result<StatusResponse, RagComputerError> {
        let path = format!("/v1/collections/{}", urlencode(name));
        self.client.transport.delete(&path).await
    }

    /// Get collection statistics.
    pub async fn stats(&self, name: &str) -> Result<CollectionStatsResponse, RagComputerError> {
        let path = format!("/v1/collections/{}/stats", urlencode(name));
        self.client.transport.get(&path, vec![]).await
    }

    /// Truncate a collection (delete all documents but keep the collection).
    pub async fn truncate(&self, name: &str) -> Result<StatusResponse, RagComputerError> {
        let path = format!("/v1/collections/{}/truncate", urlencode(name));
        self.client
            .transport
            .post(&path, &serde_json::Value::Null)
            .await
    }

    /// Queue all ready/failed documents in a collection for re-embedding.
    pub async fn reembed(&self, name: &str) -> Result<StatusResponse, RagComputerError> {
        let path = format!("/v1/collections/{}/reembed", urlencode(name));
        self.client
            .transport
            .post(&path, &serde_json::Value::Null)
            .await
    }

    /// Stream real-time events for a collection via SSE.
    pub async fn stream_events(&self, name: &str) -> Result<SseStream, RagComputerError> {
        let path = format!("/v1/collections/{}/events", urlencode(name));
        let response = self.client.transport.get_stream(&path).await?;
        Ok(SseStream::new(response))
    }
}
